### ISSUE TITLE
Editors on OSX result in random events

### ISSUE THREAD

**Issue by robfig** [howeyc/fsnotify#91]
I apologize ahead of time for this bug report.  

I'm using fsnotify to watch templates during development as a trigger to reload them into the program.  However, for whatever reason, I get a miscellaneous mix of events fired on save.  It's always one of MODIFY, MODIFY|ATTRIB, or RENAME, and sometimes it is combinations of them, or multiple of one.  

For example, repeatedly adding a byte to a file with Emacs and saving the file, results in these events being fired (separated to denote batches related to one save): 

```
"/Users/robfig/.../views/features.soy": MODIFY|ATTRIB
--
"/Users/robfig/.../views/features.soy": MODIFY|ATTRIB
--
"/Users/robfig/.../views/features.soy": MODIFY|ATTRIB
"/Users/robfig/.../views/features.soy": MODIFY
"/Users/robfig/.../views/features.soy": MODIFY
--
"/Users/robfig/.../views/features.soy": MODIFY|ATTRIB
"/Users/robfig/.../views/features.soy": MODIFY
--
"/Users/robfig/.../views/features.soy": MODIFY|ATTRIB
```

Notes:
- Most commonly, there is a single MODIFY|ATTRIB event on save.
- After a RENAME appears, fsnotify stops watching the file so no more events come out. 
- Frequently, exactly 1 second after the save, another MODIFY|ATTRIB event is fired.

I am working around this behavior with these techniques:
- Treat MODIFY|ATTRIB as MODIFY
- Drain the Event channel after each event, to avoid duplicating work on multi-events.
- Re-adding the watch if the event is a RENAME, after sleeping for 10ms.  (adding the watch immediately results in an error: `lstat /Users/robfig/.../testfile.soy: no such file or directory`)

I think this results in decent behavior for me, so this problem is not urgent from my point of view.  But since it's going into Go standard library (with the specific goal of watching for changes to code), I thought that it would be useful for me to file an issue.

Debugging notes:
- fsnotify is not doing anything obviously incorrect -- it accurately reports the events coming out kqueue. 
- I tried to write a program to reproduce this behavior, and I couldn't.  So, I suppose it has something to do with how editors interact with OSX filesystem.  However, it happens both for me (Mavericks, Emacs) and my coworker (Lion, Sublime), so I don't think it's too specific to my setup.
- DTrace (e.g. iosnoop, etc) suggested everything was fine.  Saving resulted in one "Write" DTrace event. 

I can reproduce by `NewWatcher().Watch(FILENAME)` and editing that file with my editor while printing events from that watcher in the obvious way.

I'm not really sure how to debug it further.  Happy to investigate more if anyone has ideas.

**Comment by ghost** [comment]
I don't think this should be called a bug. If a text editor (for example), does a file saving process that looks like this:
1. Write to temp file
2. Move old file out of the way
3. Move temp file onto new file

Then you should rightfully expect that there are multiple notifications yes?

By the way, there are sound reasons for doing this. If you simply overwrite a file in place, there is a tiny chance something could read your file when you are halfway through re-writing data into that file. This is especially a problem with a larger file. A good example is a web server, if you are writing a 100k text file, and a user is trying to load that text file over http at the same time, what will the user see? sure it might only take 0.1s to write that 100k text file, but what if you are serving 1000 hits per second on that page?
